src/legacy/core_plugins/kibana/public/discover/np_ready/angular/discover_state.ts

```diff
@@ -0,0 +1,268 @@
+  /**
+   * Time filter
+   */
+  time?: { from: string; to: string };
+}
+
+interface GetStateParams {
+  /**
+   * Default state used for merging with with URL state to get the initial state
+   */
+  defaultAppState?: AppState;
+  /**
+   * Determins the use of long vs. short/hashed urls
+   */
+  storeInSessionStorage?: boolean;
+  /**
+   * Browser history used for testing
+   */
+  hashHistory?: History;
+}
+
+export interface GetStateReturn {
+  /**
+   * Global state, the _g part of the URL
+   */
+  globalStateContainer: ReduxLikeStateContainer<GlobalState>;
+  /**
+   * App state, the _a part of the URL
+   */
+  appStateContainer: ReduxLikeStateContainer<AppState>;
+  /**
+   * Start sync between state and URL
+   */
+  startSync: () => void;
+  /**
+   * Stop sync between state and URL
+   */
+  stopSync: () => void;
+  /**
+   * Set app state to with a partial new app state
+   */
+  setAppState: (newState: Partial<AppState>) => void;
+  /**
+   * Set global state to with a partial new app state
+   */
+  setGlobalState: (newState: Partial<GlobalState>) => void;
+  /**
+   * Get global filters
+   */
+  getGlobalFilters: () => Filter[];
+  /**
+   * Get global filters
+   */
+  getAppFilters: () => Filter[];
+  /**
+   * Sync state to URL, used for testing
+   */
+  flushToUrl: () => void;
+  /**
+   * Reset initial state to the current app state
+   */
+  resetInitialAppState: () => void;
+  /**
+   * Returns whether the current app state is different to the initial state
+   */
+  isAppStateDirty: () => void;
+  /**
+   * Replace the current URL with the current state without adding another browser history entry
+   */
+  replaceUrlState: () => Promise<void>;
+}
+const GLOBAL_STATE_URL_KEY = '_g';
+const APP_STATE_URL_KEY = '_a';
+
+/**
+ * Builds and returns appState and globalState containers and helper functions
+ * Used to sync URL with UI state
+ */
+export function getState({
+  defaultAppState = {},
+  storeInSessionStorage = false,
+  hashHistory,
+}: GetStateParams): GetStateReturn {
+  const stateStorage = createKbnUrlStateStorage({
+    useHash: storeInSessionStorage,
+    history: hashHistory ? hashHistory : createHashHistory(),
+  });
+
+  const globalStateInitial = stateStorage.get(GLOBAL_STATE_URL_KEY) as GlobalState;
+  const globalStateContainer = createStateContainer<GlobalState>(globalStateInitial);
+
+  const appStateFromUrl = stateStorage.get(APP_STATE_URL_KEY) as AppState;
+  let initialAppState = {
+    ...defaultAppState,
+    ...appStateFromUrl,
+  };
+
+  const appStateContainer = createStateContainer<AppState>(initialAppState);
+
+  const { start, stop } = syncStates([
+    {
+      storageKey: APP_STATE_URL_KEY,
+      stateContainer: {
+        ...appStateContainer,
+        // handle null value when url switch doesn't contain state info
+        ...{ set: value => value && appStateContainer.set(value) },
+      },
+      stateStorage,
+    },
+    {
+      storageKey: GLOBAL_STATE_URL_KEY,
```

Comment:
@kertal, wonder if you saw how `_g` part is handled in dashboard. As all of the state from `_g` belongs to data plugin, there is a helper function to bind data plugin to  `_g` syncing.

https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/kibana/public/dashboard/np_ready/dashboard_app_controller.tsx#L130

Just wanted to point that one out. Up to you to decide if it also makes sense in discover.

Reply:
thx! I'm having a closer look at this way today!

Reply:
makes much sense to do it that way, just pushed an adaption, let's see if also Jenkis agrees

Reply:
should be fine now

Reply:
also the  hashed and expanded url format switch works now correctly